Application/EdFi.Admin.DataAccess/Repositories/ClientAppRepo.cs

```diff
@@ -80,6 +80,62 @@ private Profile GetOrCreateProfile(string profileName)
             }
         }
 
+        private OwnershipToken GetOrCreateOwnershipToken(string ownershipToken)
+        {
+            using (var context = _contextFactory.CreateContext())
+            {
+                var ownershipTokens = context.OwnershipTokens.FirstOrDefault(s => s.Description == ownershipToken);
+
+                if (ownershipTokens == null)
+                {
+                    context.OwnershipTokens.Add(new OwnershipToken { Description = ownershipToken });
+                    context.SaveChanges();
+                }
+
+                return context.OwnershipTokens.FirstOrDefault(s => s.Description == ownershipToken);
+            }
+        }
+
+        public void AddOwnershipTokensToApiClient(string ownershipToken, int apiClientId)
+        {
+            using (var context = _contextFactory.CreateContext())
+            {
+                var ownershiptoken = GetOrCreateOwnershipToken(ownershipToken);
+
+                var currentOwnershipToken = context.OwnershipTokens
+                    .Include(u => u.ApiClients)
+                    .FirstOrDefault(u => u.OwnershipTokenId == ownershiptoken.OwnershipTokenId);
+
+                if (!currentOwnershipToken.ApiClients.Any(a => a.ApiClientId == apiClientId))
+                {
+                    var apiClient = context.Clients.FirstOrDefault(a => a.ApiClientId == apiClientId);
+                    currentOwnershipToken.ApiClients.Add(apiClient);
+                }
+
+                context.SaveChanges();
+            }
+        }
+
+        public void AddApiClientOwnershipTokens(List<string> ownershipTokens, int apiClientId)
+        {
+            using (var context = _contextFactory.CreateContext())
+            {
+                var apiClientOwnershipTokenList = new List<ApiClientOwnershipToken>();
+                foreach (var ownershipToken in ownershipTokens)
+                {
+                    var ownershiptoken = context.OwnershipTokens.FirstOrDefault(x => x.Description == ownershipToken);
+                    var apiClient = context.Clients.FirstOrDefault(u => u.ApiClientId == apiClientId);
```

Comment:
Since _apiClientId_ doesn't change, this query could have been placed before the foreach, saving trips to the DB.